src/rdkafka_idempotence.c

```diff
@@ -0,0 +1,431 @@
+                        int all_cnt = rd_atomic32_get(&rk->rk_broker_cnt);
+                        int err_unsupported =
+                                up_cnt > 0 &&
+                                rd_interval(&rk->rk_suppress.no_idemp_brokers,
+                                            5*60*1000000/*5 minutes*/, 0) > 0;
+
+                        rd_kafka_wrunlock(rk);
+                        rd_kafka_idemp_restart_request_pid_tmr(rk, 500);
+
+                        if (err_unsupported)
+                                rd_kafka_op_err(
+                                        rk,
+                                        RD_KAFKA_RESP_ERR__UNSUPPORTED_FEATURE,
+                                        "Idempotent Producer not supported by "
+                                        "any of the %d broker(s) in state UP: "
+                                        "requires broker version >= 0.11.0",
+                                        up_cnt);
+                        else if (up_cnt == 0)
+                                rd_kafka_dbg(rk, EOS, "PIDBROKER",
+                                             "No brokers available for "
+                                             "acquiring Producer ID: "
+                                             "no brokers are up");
+                        else
+                                rd_kafka_dbg(rk, EOS, "PIDBROKER",
+                                             "None of the %d/%d brokers in "
+                                             "state UP supports "
+                                             "the Idempotent Producer: "
+                                             "requires broker "
+                                             "version >= 0.11.0",
+                                             up_cnt, all_cnt);
+                        return 0;
+                }
+        } else {
+                /* Increase passed broker's refcount so we don't
+                 * have to check if rkb should be destroyed or not below
+                 * (broker_any_usable() returns a new reference). */
+                rd_kafka_broker_keep(rkb);
+        }
+
+        rd_rkb_dbg(rkb, EOS, "GETPID", "Acquiring ProducerId: %s", reason);
+
+        err = rd_kafka_InitProducerIdRequest(
+                rkb, NULL, -1,
+                errstr, sizeof(errstr),
+                RD_KAFKA_REPLYQ(rk->rk_ops, 0),
+                rd_kafka_handle_InitProducerId, NULL);
+
+        if (!err) {
+                rd_kafka_idemp_set_state(rkb->rkb_rk,
+                                         RD_KAFKA_IDEMP_STATE_WAIT_PID);
+                rd_kafka_wrunlock(rkb->rkb_rk);
+                rd_kafka_broker_destroy(rkb);
+                return 1;
+        }
+
+        rd_kafka_wrunlock(rkb->rkb_rk);
+
+        rd_rkb_dbg(rkb, EOS, "GETPID",
+                   "Can't acquire ProducerId from this broker: %s", errstr);
+        rd_kafka_idemp_restart_request_pid_tmr(rk, 500);
```

Comment:
nit: the 500's a bit magical, used in a few places, could be a constant somewhere.